### DISCUSSION TITLE
allow both windows and linux containers at the same time on docker desktop for windows

### DISCUSSION THREAD

**Discussion by LewisPringle** [docker/roadmap#568] [Q&A]
Currently you need to track your current mode and 'swithc containers mode' which is slow and prevents using both windows and linux containers at the same time. This is very inconvenient for those who work with both windows and linux (which is probably most people using docker windows).

Seems like it should be easy too. Even if you have differnt backend processes for windows and linux, just have an intermediary process that multiplexes comms to the two (windows/linux) backends.

**Comment by kevmeister68** [comment]
I thought of this also but it’s a double-edged sword, because you presumably need a command-line switch to indicate which docker engine you wish to send the command to. This might work OK for very ad-hoc sets of commands, but if you need to perform several docker commands in a row for one docker engine, then switching engines becomes the easier option. I’m presuming use of command-line here not the Docker Desktop GUI. The GUI could benefit, however, from a highly visible widget (eg. top-right where the settings, help, account etc icons are located), rather than only in the system tray menu.

My approach is that I have two batch files, dockwin.baat, and docklin.bat, which invokes the docker-cli program with the appropriate -SwitchXXXEngine switch. Easy enough to use from command-line or call from batch script.

I have not found switching to be slow using this approach.

**Comment by samrato** [comment]
From a technical perspective, supporting Windows and Linux containers simultaneously in Docker Desktop is more complex than it initially appears because they rely on different runtimes and isolation mechanisms.

Linux containers run through WSL2 or a Linux VM, while Windows containers rely on the Windows container runtime and Windows kernel features. Although Docker Desktop could theoretically expose both engines at once, the challenge is providing a consistent user experience, API endpoint, networking model, volume mounts, and CLI context management across two active container engines.

One possible approach would be to expose each engine as a separate Docker context and allow users to switch contexts without changing the active backend. This would make it possible to manage Linux and Windows workloads concurrently while keeping commands explicit.

For now, using separate Docker contexts, dedicated scripts, or the current engine-switching mechanism is likely the most practical solution. It would be interesting to hear whether the Docker team has evaluated running both engines simultaneously and what the main architectural blockers are.
